streams/src/main/java/org/apache/kafka/streams/TopologyConfig.java

```diff
@@ -216,12 +232,38 @@ public TopologyConfig(final String topologyName, final StreamsConfig globalAppCo
         } else {
             storeType = globalAppConfigs.getString(DEFAULT_DSL_STORE_CONFIG);
         }
+
+        if (isTopologyOverride(DSL_STORE_SUPPLIERS_CLASS_CONFIG, topologyOverrides)) {
+            dslStoreSuppliers = getString(DSL_STORE_SUPPLIERS_CLASS_CONFIG);
+            log.info("Topology {} is overriding {} to {}", topologyName, DSL_STORE_SUPPLIERS_CLASS_CONFIG, dslStoreSuppliers);
+        } else {
+            dslStoreSuppliers = globalAppConfigs.getString(DSL_STORE_SUPPLIERS_CLASS_CONFIG);
+        }
     }
 
+    @Deprecated
     public Materialized.StoreType parseStoreType() {
         return MaterializedInternal.parse(storeType);
     }
 
+    /**
+     * @return the DslStoreSuppliers if the value was explicitly configured (either by
+     *         {@link StreamsConfig#DEFAULT_DSL_STORE} or {@link StreamsConfig#DSL_STORE_SUPPLIERS_CLASS_CONFIG})
+     */
+    public Optional<DslStoreSuppliers> resolveDslStoreSuppliers() {
```

Comment:
@ableegoldman - I'll add this back to the KIP, LMK if you think this API makes sense since it is technically public

Reply:
Does this method even need to be in the `TopologyConfig` class? Seems like we should be able to do everything here from some static utility, in which case we don't need this.

We definitely have a real problem with mixing up utility methods that are meant for internal use only with the by-definition public interface of the XXXConfig API. The Config classes just have some real fundamental abstractions issues. But we should try not to make the problem any worse at least, if we can help it.

(In fact Matthias has a PR/draft KIP to clean up StreamsConfig: https://github.com/apache/kafka/pull/14548)

Reply:
we can, the alternative though is having the `dslStoreSuppliers` field be nullable -- I'm not sure what's prefereable.

Reply:
> having the dslStoreSuppliers field be nullable 

This is technically a different part of the code but on the subject of it being nullable... see https://github.com/apache/kafka/pull/14648#discussion_r1396543979

Anyways I personally feel like it would be preferable to expose the globalAppConfigs over the entire thing, or better yet, something like this which at least kind of makes sense as a generally-useful public API:
```
public boolean isGlobalOverride(final String config) {
    return globalAppconfigs.originals().containsKey(config);
}
```
Basically something that mirrors the `isTopologyOverride`, which presumably we also need. Just a thought.

Reply:
note that `isTopologyOverride` is a private helper method, it's not exposed. :/ looks like none of the options are particularly good here

Reply:
Ah....in that case I would say just do whatever is easiest and cleanest for now (so presumably, just making this public)

If this class was intended to be a real and lasting public interface then it's probably neater to just have `isTopologyOverride` and `isGlobalOverride` as public APIs and then do the actual DslStoreSupplier resolution in an internal utility class. But I think we'll want to just deprecate/remove this entire class soon anyways, so it doesn't really hurt to have one more ugly API if we consider it effectively temporary. Maybe Matthias can even tackle this (deprecating TopologyConfig) in the KIP he was planning to clean up similar ugly APIs over in StreamsConfig -- I'll bring it up with him.

FYI though it's probably worth defending this choice when you update the KIP thread, with some form of the above argument. It's definitely not a good public interface but the whole thing needs to go and we can deprecate it all together

Reply:
I'm a bit inclined with adding such an `resolveDslStoreSuppliers` to `StreamsConfig` directly as long as we label it EVOLVING to leave us options to delete it (of course this would need get other's opinion as we update the KIP DISCUSS thread), until we have a clear clue on what we want to expose eventually.